controllers/observability_reconciler.go

```diff
@@ -0,0 +1,420 @@
+	TypeMeta: metav1.TypeMeta{
+		Kind:       monitoringv1.PodMonitorsKind,
+		APIVersion: monitoringv1.SchemeGroupVersion.String(),
+	},
+	ObjectMeta: metav1.ObjectMeta{
+		Name: envoyStatsMonitorName,
+		Labels: map[string]string{
+			kuadrantObservabilityLabel: "true",
+		},
+	},
+	Spec: monitoringv1.PodMonitorSpec{
+		PodMetricsEndpoints: []monitoringv1.PodMetricsEndpoint{{
+			Port: "http-envoy-prom",
+			Path: "/stats/prometheus",
+		}},
+		Selector: metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"app": "kuadrant-ingressgateway",
+			},
+		},
+	},
+}
+
+type ObservabilityReconciler struct {
+	Client     *dynamic.DynamicClient
+	restMapper meta.RESTMapper
+}
+
+func NewObservabilityReconciler(client *dynamic.DynamicClient, rm meta.RESTMapper) *ObservabilityReconciler {
+	return &ObservabilityReconciler{
+		Client:     client,
+		restMapper: rm,
+	}
+}
+
+func (r *ObservabilityReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{ReconcileFunc: r.Reconcile, Events: []controller.ResourceEventMatcher{
+		{Kind: &kuadrantv1beta1.KuadrantGroupKind},
+		{Kind: ptr.To(schema.GroupKind{Group: monitoringv1.SchemeGroupVersion.Group, Kind: monitoringv1.ServiceMonitorsKind}), EventType: ptr.To(controller.DeleteEvent)},
+		{Kind: ptr.To(schema.GroupKind{Group: monitoringv1.SchemeGroupVersion.Group, Kind: monitoringv1.PodMonitorsKind}), EventType: ptr.To(controller.DeleteEvent)},
+		{Kind: &machinery.GatewayGroupKind}, // all events
+		{Kind: &machinery.GatewayClassGroupKind, EventType: ptr.To(controller.CreateEvent)},
+		{Kind: &machinery.GatewayClassGroupKind, EventType: ptr.To(controller.DeleteEvent)},
+	},
+	}
+}
+
+func (r *ObservabilityReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, _ *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("ObservabilityReconciler")
+	logger.V(1).Info("reconciling observability", "status", "started")
+	defer logger.V(1).Info("reconciling observability", "status", "completed")
+
+	// Get all monitors first, if any exist
+	monitorObjs := lo.FilterMap(topology.Objects().Items(), func(item machinery.Object, _ int) (machinery.Object, bool) {
+		if item.GroupVersionKind().Kind == monitoringv1.ServiceMonitorsKind || item.GroupVersionKind().Kind == monitoringv1.PodMonitorsKind {
+			return item, true
+		}
+		return nil, false
+	})
+
+	// Check that a kuadrant resource exists, and observability enabled,
+	// otherwise delete all monitors
+	kObj := GetKuadrantFromTopology(topology)
+	if kObj == nil || !kObj.Spec.Observability.Enable {
+		logger.V(1).Info("deleting any existing monitors", "kuadrant", kObj != nil)
+		r.deleteAllMonitors(ctx, monitorObjs, logger)
+		return nil
+	}
+
+	// Create all monitors
+	logger.V(1).Info("observability enabled, creating monitors")
+
+	// Kuadrant Operator monitor
+	r.createMonitor(ctx, monitorObjs, kOpMonitorSpec, kObj.Namespace, logger)
```

Comment:
check my comment about namespaces. `OPERATOR_NAMESPACE` env var could be a better option. It can be used via global variable `operatorNamespace` in this `controllers` package.

Reply:
Same for the other operators. I think we can make a "safe" assumption (this will shoot back in our face sooner or later) that all operators are being installed in the same namespace as the kuadrant operator.